File: src/domain/preferences.ts
export const themeModes = ["system", "light", "dark"] as const;

export type ThemeMode = (typeof themeModes)[number];
export type WindowBehavior = "hide_to_tray";
export const windowPresets = ["sticky", "iphone5", "pocket", "book", "custom"] as const;
export type WindowPreset = (typeof windowPresets)[number];
export const stickyModes = ["compact", "mini"] as const;
export type StickyMode = (typeof stickyModes)[number];
export const readerSkins = ["grid"] as const;
export type ReaderSkin = (typeof readerSkins)[number];
export const readerFontSizes = ["small", "standard", "large"] as const;
export type ReaderFontSize = (typeof readerFontSizes)[number];
export const readerLineSpacings = ["compact", "standard", "relaxed"] as const;
export type ReaderLineSpacing = (typeof readerLineSpacings)[number];

export type StickyPosition = {
  xRatio: number;
  yRatio: number;
  snap: "top_left" | "top_right" | "bottom_left" | "bottom_right" | null;
};

export type Preferences = {
  schemaVersion: 1;
  theme: ThemeMode;
  alwaysOnTop: boolean;
  windowBehavior: WindowBehavior;
  windowPreset: WindowPreset;
  stickyPosition: StickyPosition;
  stickyMode: StickyMode;
  readerSkin: ReaderSkin;
  readerFontSize: ReaderFontSize;
  readerLineSpacing: ReaderLineSpacing;
  readerContentVisible: boolean;
  currentReaderDocumentId: string | null;
};

export const defaultPreferences: Preferences = {
  schemaVersion: 1,
  theme: "system",
  alwaysOnTop: false,
  windowBehavior: "hide_to_tray",
  windowPreset: "sticky",
  stickyPosition: { xRatio: 0.5, yRatio: 0.28, snap: null },
  stickyMode: "compact",
  readerSkin: "grid",
  readerFontSize: "standard",
  readerLineSpacing: "standard",
  readerContentVisible: true,
  currentReaderDocumentId: null,
};

export type CompactGeometry = {
  boardWidth: number;
  boardHeight: number;
  stickyWidth: number;
  stickyHeight: number;
  left: number;
  top: number;
};

export type DragFrame = Pick<
  CompactGeometry,
  "boardWidth" | "boardHeight" | "stickyWidth" | "stickyHeight"
> & {
  originLeft: number;
  originTop: number;
  deltaX: number;
  deltaY: number;
};

export function compactDragFrame(frame: DragFrame) {
  return {
    left: Math.max(
      12,
      Math.min(frame.boardWidth - frame.stickyWidth - 12, frame.originLeft + frame.deltaX),
    ),
    top: Math.max(
      76,
      Math.min(frame.boardHeight - frame.stickyHeight - 12, frame.originTop + frame.deltaY),
    ),
  };
}

export function settleCompactPosition(
  geometry: CompactGeometry,
  snapDistance = 28,
): StickyPosition {
  const { boardWidth, boardHeight, stickyWidth, stickyHeight } = geometry;
  const left = Math.max(12, Math.min(boardWidth - stickyWidth - 12, geometry.left));
  const top = Math.max(76, Math.min(boardHeight - stickyHeight - 12, geometry.top));
  const horizontal =
    left < snapDistance ? "left" : boardWidth - stickyWidth - left < snapDistance ? "right" : null;
  const vertical =
    top - 76 < snapDistance
      ? "top"
      : boardHeight - stickyHeight - top < snapDistance
        ? "bottom"
        : null;
  const snap =
    horizontal && vertical ? (`${vertical}_${horizontal}` as StickyPosition["snap"]) : null;
  const anchors = {
    top_left: { left: 12, top: 76 },
    top_right: { left: boardWidth - stickyWidth - 12, top: 76 },
    bottom_left: { left: 12, top: boardHeight - stickyHeight - 12 },
    bottom_right: { left: boardWidth - stickyWidth - 12, top: boardHeight - stickyHeight - 12 },
  };
  const settled = snap ? anchors[snap] : { left, top };
  return {
    xRatio: Math.max(0, Math.min(1, (settled.left + stickyWidth / 2) / boardWidth)),
    yRatio: Math.max(0, Math.min(1, (settled.top + stickyHeight / 2) / boardHeight)),
    snap,
  };
}

File: src-tauri/src/domain/preferences.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ThemeMode {
    #[default]
    System,
    Light,
    Dark,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WindowBehavior {
    #[default]
    HideToTray,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WindowPreset {
    #[default]
    Sticky,
    Iphone5,
    Pocket,
    Book,
    Custom,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StickyMode {
    #[default]
    Compact,
    Mini,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReaderSkin {
    #[default]
    Grid,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReaderFontSize {
    Small,
    #[default]
    Standard,
    Large,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReaderLineSpacing {
    Compact,
    #[default]
    Standard,
    Relaxed,
}

impl WindowPreset {
    pub const fn logical_size(self) -> Option<(f64, f64)> {
        match self {
            Self::Sticky => Some((320.0, 420.0)),
            Self::Iphone5 => Some((320.0, 568.0)),
            Self::Pocket => Some((360.0, 640.0)),
            Self::Book => Some((420.0, 594.0)),
            Self::Custom => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StickyPosition {
    pub x_ratio: f64,
    pub y_ratio: f64,
    pub snap: Option<StickySnap>,
}

impl Eq for StickyPosition {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StickySnap {
    TopLeft,
    TopRight,
    BottomLeft,
    BottomRight,
}

impl Default for StickyPosition {
    fn default() -> Self {
        Self {
            x_ratio: 0.5,
            y_ratio: 0.28,
            snap: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Preferences {
    pub schema_version: u8,
    pub theme: ThemeMode,
    pub always_on_top: bool,
    pub window_behavior: WindowBehavior,
    #[serde(default)]
    pub window_preset: WindowPreset,
    #[serde(default)]
    pub sticky_position: StickyPosition,
    #[serde(default)]
    pub sticky_mode: StickyMode,
    #[serde(default)]
    pub reader_skin: ReaderSkin,
    #[serde(default)]
    pub reader_font_size: ReaderFontSize,
    #[serde(default)]
    pub reader_line_spacing: ReaderLineSpacing,
}

impl Default for Preferences {
    fn default() -> Self {
        Self {
            schema_version: 1,
            theme: ThemeMode::System,
            always_on_top: false,
            window_behavior: WindowBehavior::HideToTray,
            window_preset: WindowPreset::Sticky,
            sticky_position: StickyPosition::default(),
            sticky_mode: StickyMode::Compact,
            reader_skin: ReaderSkin::Grid,
            reader_font_size: ReaderFontSize::Standard,
            reader_line_spacing: ReaderLineSpacing::Standard,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serializes_the_frontend_wire_contract() {
        let value = serde_json::to_value(Preferences::default()).expect("serialize preferences");

        assert_eq!(value["schemaVersion"], 1);
        assert_eq!(value["theme"], "system");
        assert_eq!(value["alwaysOnTop"], false);
        assert_eq!(value["windowBehavior"], "hide_to_tray");
        assert_eq!(value["windowPreset"], "sticky");
        assert_eq!(value["stickyPosition"]["xRatio"], 0.5);
        assert_eq!(value["stickyMode"], "compact");
        assert_eq!(value["readerSkin"], "grid");
        assert_eq!(value["readerFontSize"], "standard");
        assert_eq!(value["readerLineSpacing"], "standard");
    }

    #[test]
    fn defaults_legacy_preferences_to_compact_mode() {
        let preferences: Preferences = serde_json::from_value(serde_json::json!({
            "schemaVersion": 1,
            "theme": "light",
            "alwaysOnTop": false,
            "windowBehavior": "hide_to_tray",
            "windowPreset": "book",
            "stickyPosition": { "xRatio": 0.7, "yRatio": 0.4, "snap": null }
        }))
        .expect("deserialize legacy preferences");

        assert_eq!(preferences.sticky_mode, StickyMode::Compact);
        assert_eq!(preferences.reader_skin, ReaderSkin::Grid);
        assert_eq!(preferences.reader_font_size, ReaderFontSize::Standard);
        assert_eq!(preferences.reader_line_spacing, ReaderLineSpacing::Standard);
    }
}
